### ISSUE TITLE
Handle mail data of projects that use multiple mailing lists correctly

### ISSUE THREAD

**Issue by bockthom** [se-sic/coronet#266]
### Description

Coronet is able to process e-mails from multiple mailing lists within a project. However, up until now, we did not take care of e-mails that belong to multiple mailing lists.

Example:

|Message-Id | Date | Thread Id|
|---|---|---|
| 1 | 2020-05-02 17:30 | 1#25 |
| 2 | 2024-04-04 14:44 | 2#234 |

In the above example, we can see that we have two different e-mails. E-mail 1 was sent to thread 25 of mailing list 1.
E-mail 2 was sent to thread 234 of mailing list 2. This is what coronet can handle properly.

But let's assume the following use case:

E-mail 3 was sent to both mailing lists 1 and 2. So we have the same message id with multiple entries:

|Message-Id | Date | Thread Id|
|---|---|---|
| 3 | 2015-15-15 15:15 | 1#89 |
| 3 | 2015-15-15 15:15 | 2#17 |

In this case, e-mail 3 occurs twice in the mail data - once with its thread id on mailing list 1, and once with its thread id on mailing list 2. (This is necessary to be able to construct mail networks properly. Having one entry per mail and a comma-separated list of thread-ids would make network construction much more complex.)

This is a situation which we have not thought about when modeling mail data into coronet. Up until now, I have assumed that every e-mail is only contained once (i.e., with one line) in the mail data. But as the example above shoes, there can also be multiple entries for each e-mail. Therefore, we need to make sure that this case is properly handled. That is, we need to make sure that we count each e-mail only once when computing count-based statistics and generating lists of message ids etc. 

(The situation is similar to how we handle commit data: There, we also have multiple entries for each commit in the commit data if multiple files/functions are affected by the commit. I guess that we need to apply similar unification approaches for getting unique message ids from the mail data as we did for the extraction of unique commit hashes.)

### Suggestion
I suggest to check any and all parts of coronet in which mail data are handled, whether they deal with multiple entries as expected (similar to how multiple entries for commit data are handled).

**Comment by MaLoefUDS** [comment]
Hey, I have been searching coronet for occurrences of `"mails"` and `"count"` to find places in which such mails may be causing issues.

The main usage spots I found were `util-networks-covariates.R` where attributes can be attached to networks based on statistical features. These functions then call helpers for the statistic evaluating that can be found in `util-data-misc.R` (l. 326-558). In all of the relevant functions, multi-thread-mails are treated correctly from what I understand of it.

What I have also checked is creating activity-based bins for splitting. I found that since the `message.id` is identical for when a mail is sent to multiple threads, therefore the splitting does treat the mails all as one basically and puts them all into one bin just as it does with consecutive cochange artifacts that share the same id aka commit hash. I believe there is no issue here. 

Another interesting spot is in `util-data.R` in the function `filter.patchstack.mails`. Do you also want to remove mails that occur in a patchstack from other threads? Say a mail is in `thread#1` and `thread#2`. In `thread#2` it is in a patchstack. Do you want to remove it from `thread#1` as well? This is currently not the case.

**Comment by bockthom** [comment]
> Another interesting spot is in `util-data.R` in the function `filter.patchstack.mails`. Do you also want to remove mails that occur in a patchstack from other threads? Say a mail is in `thread#1` and `thread#2`. In `thread#2` it is in a patchstack. Do you want to remove it from `thread#1` as well? This is currently not the case.

This is an interesting question. I tend not to remove them in other threads, but I am not sure yet. Let's discuss the following example:

|Message-Id | Date | Thread Id| Subject | Author |
|---|---|---|---|---|
| 1 | 2020-05-02 17:30:00 | 1#25 | [PATCH 00/02] cover letter | Olaf |
| 2 | 2020-05-02 17:30:02 | 1#25 | [PATCH 02/02] first patch | Olaf |
| 3 | 2020-05-02 17:30:04 | 1#25 | [PATCH 02/02] second patch | Olaf |
| 3 | 2020-05-02 17:30:04 | 2#30 | [PATCH 02/02] second patch | Olaf |
| 4 | 2020-05-02 21:35:00 | 2#30 | Re: [PATCH 02/02] second patch | Claus |
| 5 | 2020-05-02 22:00:13 | 2#30 | Re: [PATCH 02/02] second patch | Olaf |

When thread `1#25` really is a patchstack, a situation such as in this example is **not possible**: Mails that form a patchstack always are automatically generated and belong to the same thread. So, it is not possible that [PATCH 02/02] belongs to a thread that [PATCH 00/02] does not belong to. Also answers to this e-mail belong to the initial thread as they use one of the patchstack mails' `In-Reply-To:` or `References:` header fields for thread detection.

If the entire patchstack is sent to multiple mailing lists, all patches occur at all lists, and therefore patchstack-filtering should be automatically performed on all mailing lists (since each mailing list contains the complete patchstack). Could you please check that again?

The only thing that is possible is that either one of the mails is forwarded to another mailing list or the reply to a patchstack mail is sent to another mailing list – but this is not a problem for patchstack filtering, as the forward or reply mails are not part of the patchstack itself.

So, I believe there is no situation where individual mails of a patchstack are sent to multiple threads and therefore we should not have any problems with duplicated message ids during patchstack filtering.

**Comment by MaLoefUDS** [comment]
I have checked that if two mail threads both have the same patchstack of mails, that it will be removed from both. Also I generally agree with your assessment and do not think that this there is any issue here.

**Comment by bockthom** [comment]
Thanks for checking. As we agree that handling message ids that occur on multiple mailing lists does not require any changes, I will close this issue.
